src/client/keycode.cpp

```diff
@@ -221,122 +223,172 @@ static const struct table_key table[] = {
+		lookup_keykey(std::get<irr::EKEY_CODE>(scancode)) :
+		lookup_scancode(std::get<u32>(scancode));
+}
+
+void KeyPress::loadFromKey(irr::EKEY_CODE keycode, wchar_t keychar)
+{
+	if (auto device = RenderingEngine::get_raw_device()) {
+		scancode = device->getScancodeFromKey(Keycode(keycode, keychar));
 	} else {
-		// Lookup by name
-		m_name = name;
-		try {
-			auto &k = lookup_keyname(name);
-			Key = k.Key;
-			Char = k.Char;
-			return;
-		} catch (UnknownKeycode &e) {};
+		warningstream << "IrrlichtDevice not available when creating KeyPress." <<
+				" The resulting value may be incorrect or unusable." << std::endl;
+		if (Keycode::isValid(keycode))
+			scancode = keycode;
+		else
+			scancode = (u32) keychar;
 	}
+}
 
-	// It's not a known key, complain and try to do something
-	Key = irr::KEY_KEY_CODES_COUNT;
-	int chars_read = mbtowc(&Char, name, 1);
-	FATAL_ERROR_IF(chars_read != 1, "Unexpected multibyte character");
-	m_name = "";
-	warningstream << "KeyPress: Unknown key '" << name
-		<< "', falling back to first char." << std::endl;
+KeyPress::KeyPress(const std::string &name)
+{
+	if (loadFromScancode(name))
+		return;
+	const auto &key = lookup_keyname(name);
+	loadFromKey(key.Key, key.Char);
 }
 
-KeyPress::KeyPress(const irr::SEvent::SKeyInput &in, bool prefer_character)
+KeyPress::KeyPress(const irr::SEvent::SKeyInput &in)
 {
-	if (prefer_character)
-		Key = irr::KEY_KEY_CODES_COUNT;
+#if USE_SDL2
+	if (in.SystemKeyCode)
+		scancode.emplace<u32>(in.SystemKeyCode);
 	else
-		Key = in.Key;
-	Char = in.Char;
+		scancode.emplace<irr::EKEY_CODE>(in.Key);
+#else
+	loadFromKey(in.Key, in.Char);
+#endif
+}
 
-	try {
-		if (valid_kcode(Key))
-			m_name = lookup_keykey(Key).Name;
-		else
-			m_name = lookup_keychar(Char).Name;
-	} catch (UnknownKeycode &e) {
-		m_name.clear();
-	};
+std::string KeyPress::formatScancode() const
+{
+#if USE_SDL2
+	if (auto pv = std::get_if<u32>(&scancode))
+		return *pv == 0 ? "" : "<" + std::to_string(*pv) + ">";
+#endif
+	return "";
 }
 
-const char *KeyPress::sym() const
+std::string KeyPress::sym() const
 {
-	return m_name.c_str();
+	std::string name = lookup_scancode(scancode).Name;
+	if (USE_SDL2 || name.empty())
+		if (auto newname = formatScancode(); !newname.empty())
+			return newname;
+	return name;
 }
 
-const char *KeyPress::name() const
+std::string KeyPress::name() const
 {
-	if (m_name.empty())
-		return "";
-	const char *ret;
-	if (valid_kcode(Key))
-		ret = lookup_keykey(Key).LangName;
-	else
-		ret = lookup_keychar(Char).LangName;
-	return ret ? ret : "<Unnamed key>";
+	const auto &name = lookup_scancode(scancode).LangName;
+	if (!name.empty())
+		return name;
+	return formatScancode();
+}
+
+irr::EKEY_CODE KeyPress::getKeycode() const
+{
+	return lookup_scancode(scancode).Key;
+}
+
+wchar_t KeyPress::getKeychar() const
+{
+	return lookup_scancode(scancode).Char;
 }
 
-const KeyPress EscapeKey("KEY_ESCAPE");
+bool KeyPress::loadFromScancode(const std::string &name)
+{
+#if USE_SDL2
```

Comment:
Same as in `formatScancode`. `#if` can be replaced by `if` to check the code during compile-time.

Reply:
https://github.com/luanti-org/luanti/pull/14964/commits/c54a626fe3142c71c770563084471344adc3e6df